File: backend/app/schemas/testcase.py
"""用例相关 Schema"""

from pydantic import BaseModel, Field
from datetime import datetime


class TestStepSchema(BaseModel):
    seq: int
    action: str
    target: str
    value: str | None = None
    expected: str
    timeout: int | None = 30000


class TestCaseCreate(BaseModel):
    project_id: int
    title: str = Field(..., min_length=1, max_length=256)
    description: str | None = None
    priority: str = "P2"
    status: str = "draft"
    preconditions: str | None = None
    steps: list[TestStepSchema] = []
    tags: list[str] = []
    module: str | None = None
    zentao_id: int | None = None
    zentao_case_id: str | None = None


class TestCaseUpdate(BaseModel):
    title: str | None = None
    description: str | None = None
    priority: str | None = None
    status: str | None = None
    preconditions: str | None = None
    steps: list[TestStepSchema] | None = None
    tags: list[str] | None = None
    module: str | None = None
    zentao_id: int | None = None
    zentao_case_id: str | None = None


class TestCaseResponse(BaseModel):
    id: int
    project_id: int
    title: str
    description: str | None = None
    priority: str
    status: str
    preconditions: str | None = None
    steps: list
    tags: list
    module: str | None = None
    zentao_id: int | None = None
    zentao_case_id: str | None = None
    created_by: int
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class AIGenerateStepsReq(BaseModel):
    description: str = Field(..., min_length=10, description="自然语言描述的测试场景")


class AIGenerateStepsRes(BaseModel):
    steps: list[TestStepSchema]

File: frontend/src/types/testcase.ts
// 用例相关类型定义

export type TestCaseStatus = 'draft' | 'reviewing' | 'approved' | 'deprecated'
export type TestCasePriority = 'P0' | 'P1' | 'P2' | 'P3'
export type TestStepAction = 'navigate' | 'click' | 'input' | 'select' | 'hover' | 'scroll' | 'wait' | 'assert' | 'screenshot' | 'ai_action'

export interface TestStep {
  seq: number
  action: TestStepAction
  target: string        // 目标元素描述（自然语言）或 CSS/XPath
  value?: string        // 输入值
  expected: string      // 预期结果
  timeout?: number      // 超时（ms）
}

export interface TestCase {
  id: number
  projectId: number
  title: string
  description: string
  priority: TestCasePriority
  status: TestCaseStatus
  preconditions?: string
  steps: TestStep[]
  tags: string[]
  module?: string       // 所属模块
  zentaoId?: number     // 关联禅道用例 ID
  zentaoCaseId?: string // 禅道用例编号
  createdBy: number
  creatorName?: string
  createdAt: string
  updatedAt: string
  lastRunAt?: string
  lastRunStatus?: 'passed' | 'failed' | 'skipped'
}

export interface TestCaseFilter {
  projectId?: number
  status?: TestCaseStatus
  priority?: TestCasePriority
  keyword?: string
  tags?: string[]
  module?: string
}
